### ISSUE TITLE
Differentiable power spectrum and bispectrum via JAX backend

### ISSUE THREAD

**Issue by sambit-giri** [sambit-giri/tools21cm#81]
## Motivation

Differentiable polyspectrum estimators are increasingly needed for:

- **Gradient-based inference** (e.g. HMC / NUTS samplers, SBI with gradient matching)
- **Neural network training** where P(k) or B(k) appear as loss terms
- **Sensitivity / Fisher forecasts** via automatic differentiation rather than finite differences

## Proposed API

Add an optional `backend` keyword to the main polyspectrum functions:

```python
# Power spectrum
Pk, ks = power_spectrum_1d(delta, kbins=20, box_dims=200.0, backend='jax')
Pk2d, kper, kpar = power_spectrum_2d(delta, kbins=20, box_dims=200.0, backend='jax')

# Bispectrum
bs = Bispectrum(delta, box_dims=200.0, backend='jax')
Bk, ks = bs.bispectrum_k(dk=0.3, n_bins=8)
```

Default remains `backend='numpy'` (current behaviour, no breaking change).

When `backend='jax'`, all internal array operations and FFTs use `jax.numpy` / `jax.numpy.fft`, making the estimators end-to-end differentiable via `jax.grad` / `jax.jacfwd`.

## Implementation sketch

1. Add a `_get_backend(backend)` helper that returns the appropriate `(np_module, fft_module)` pair:
   - `'numpy'` → `(numpy, numpy.fft)`
   - `'jax'` → `(jax.numpy, jax.numpy.fft)`

2. Replace hardcoded `np.*` and `np.fft.*` calls in `power_spectrum_nd`, `power_spectrum_1d`, `power_spectrum_2d`, and `Bispectrum` with the dispatched modules.

3. Bin-averaging (histogramming) needs special care — replace `np.histogram`/`np.digitize` with JAX-compatible scatter/segment operations (`jax.ops.segment_sum` or `jnp.zeros(n).at[idx].add(vals)`).

4. `Bispectrum` class: if `backend='jax'`, store the field as a `jax.Array` from the constructor so gradients flow through `bispectrum_k` and `bispectrum_k1k2`.

## Scope

- [x] `power_spectrum_nd` (core FFT step)
- [x] `power_spectrum_1d` (radial binning)
- [x] `power_spectrum_2d` (cylindrical binning)
- [x] `cross_power_spectrum_1d`
- [ ] `Bispectrum.bispectrum_k`
- [ ] `Bispectrum.bispectrum_k1k2`
- [ ] Unit tests: `jnp.allclose` comparison between numpy and JAX outputs; gradient smoke test via `jax.grad(lambda d: power_spectrum_1d(d, ...)[0].sum())(delta)`
- [ ] Notebook demo cell in `polyspectra_tutorial.ipynb`

## Dependencies

JAX is an optional dependency — gate the import behind `try/except ImportError` so the package remains installable without JAX.

```
pip install "tools21cm[jax]"   # installs jax as optional extra
```

**Comment by sambit-giri** [comment]
First implementation landed in 28aa232.

## What's in

`backend='numpy' | 'jax' | 'torch'` keyword added to:

- `power_spectrum_nd`, `cross_power_spectrum_nd` (FFT core)
- `radial_average` → `power_spectrum_1d`, `cross_power_spectrum_1d` (radial binning)
- `power_spectrum_2d` (cylindrical binning)

Both jax **and** torch are supported so downstream codes (e.g. BEoRN) can stay on one framework whichever backend they run — torch also covers GPU on Apple Silicon via MPS, where jax-metal is still immature.

## Design

- **Default numpy path untouched** — no behaviour change, existing tests pass.
- jax/torch paths: FFT + all array ops in the chosen framework → end-to-end differentiable (`jax.grad` / torch autograd), `jax.jit`-compatible, GPU-capable.
- **Static/differentiable split**: bin geometry (k grid, bin edges, per-bin mode counts, empty-bin fill gather indices) depends only on grid shape + `box_dims` + binning scheme, so it is computed in plain numpy and is constant under jit/autograd. Only the power values carry gradients; they are summed per bin with a differentiable scatter-add (`jnp.zeros(n).at[idx].add(vals)` / `torch.zeros(n).index_add(0, idx, vals)`) — exactly the segment-sum approach sketched above.
- `binning='log'` and `'linear'` supported for jax/torch (raise `NotImplementedError` otherwise); `adaptive`/`kde` variants remain numpy-only for now.
- `fill_empty_bins` reproduced as a static nearest-neighbour gather (1D) / EDT gather (2D), so it stays differentiable.
- jax note: enable `jax_enable_x64` for float64; default jax config runs the estimator in float32.

## Tests (`tests/test_PowerSpectrumBackends.py`, 8 new, all green)

- numpy↔jax and numpy↔torch parity to rtol 1e-8 for 1d (log + linear), cross-1d, 2d
- `jax.grad` of `Pk.sum()` w.r.t. the input field: finite, non-zero, matches a directional finite difference to rtol 1e-4
- `jax.jit` parity
- torch autograd gradient ≡ jax gradient (rtol 1e-6)

## Remaining scope

- [ ] `Bispectrum.bispectrum_k` / `bispectrum_k1k2` (`bispectrum.py`)
- [ ] `power_spectrum_mu` / `mu_binning` (same static-geometry + scatter-add pattern applies)
- [ ] `dimensionless_ps` (currently goes through `interp1d`; differentiable variant would use `jnp.interp`)
- [ ] adaptive binning for jax/torch (edges are data-independent given the grid, so it is actually static too — just needs wiring)
- [ ] notebook demo cell in `polyspectra_tutorial.ipynb`
